### PR TITLE
Port TargetIndexMatcher

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This is a straightforward port from Android.

### CODE REVIEW

File: packages/firestore/test/unit/model/target_index_matcher.test.ts [hunk 1]
```diff
@@ -0,0 +1,687 @@
...
+    );
+
+    for (const query of queriesWithEqualitiesAndDescendingOrder) {
+      validateServesTarget(query, 'a', IndexKind.ASCENDING);
+      validateDoesNotServeTarget(query, 'b', IndexKind.ASCENDING);
+      validateDoesNotServeTarget(query, 'a', IndexKind.CONTAINS);
+    }
+  });
+
+  it('inequalitiesWithDefaultOrder', () => {
```
Comment (ehsannas) [inline packages/firestore/test/unit/model/target_index_matcher.test.ts:141]: inequalities with default order ?

Same for the next two tests

File: packages/firestore/test/unit/model/target_index_matcher.test.ts [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+    validateServesTarget(q, 'a', IndexKind.ASCENDING, 'b', IndexKind.ASCENDING);
+  });
+
+  it('validates collection', () => {
+    {
+      const targetIndexMatcher = new TargetIndexMatcher(
+        queryToTarget(query('collId'))
+      );
+      const index = fieldIndex('collId');
+      expect(() => targetIndexMatcher.servedByIndex(index)).to.not.throw;
```
Comment (ehsannas) [inline packages/firestore/test/unit/model/target_index_matcher.test.ts:199]: I don't see any `throw` statements in target_index_matcher.ts. Similarly, I don't see any throw statements in TargetIndexMatcher.java.

does `.to.throw` and `.to.not.throw` apply to assertions (e.g. the `debugAssert` in the code)? The other examples of `.to.throw` that I found in other test files corresponded to a `throw` statement in the source.

File: packages/firestore/test/unit/model/target_index_matcher.test.ts [hunk 3]
```diff
@@ -0,0 +1,687 @@
...
+    }
+
+    {
+      const targetIndexMatcher = new TargetIndexMatcher(
+        queryToTarget(query('collId2'))
+      );
+      const index = fieldIndex('collId');
+      expect(() => targetIndexMatcher.servedByIndex(index)).to.throw(
+        'Collection IDs do not match'
```
Comment (ehsannas) [inline packages/firestore/test/unit/model/target_index_matcher.test.ts:216]: very well... I guess debugAssert counts.

File: packages/firestore/test/unit/model/target_index_matcher.test.ts [hunk 4]
```diff
@@ -0,0 +1,687 @@
...
+
+  it('with array-contains', () => {
+    for (const query of queriesWithArrayContains) {
+      validateDoesNotServeTarget(query, 'a', IndexKind.ASCENDING);
+      validateDoesNotServeTarget(query, 'a', IndexKind.ASCENDING);
+      validateServesTarget(query, 'a', IndexKind.CONTAINS);
+    }
+  });
+
+  it('array-contains is idempotent', () => {
```
Comment (ehsannas) [inline packages/firestore/test/unit/model/target_index_matcher.test.ts:229]: ts:idempotent
java:independent

**Review by ehsannas — COMMENTED** [review]
Looks good. Just a few nits

File: packages/firestore/src/model/target_index_matcher.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  fieldIndexGetArraySegment,
+  fieldIndexGetDirectionalSegments,
+  IndexKind,
+  IndexSegment
+} from './field_index';
+
+/**
+ * A light query planner for Firestore.
+ *
+ * This class matches a `FieldIndex`z against a Firestore Query `Target`. It
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/target_index_matcher.ts:38]: Done

File: packages/firestore/src/model/target_index_matcher.ts [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+ * where('a', '>=', 'a').orderBy('a').orderBy('b')     | a ASC
+ * where('a', 'array-contains', 'a').orderBy('b')      | a CONTAINS, b ASCENDING
+ * where('a', 'array-contains', 'a').orderBy('b')      | a CONTAINS
+ */
+export class TargetIndexMatcher {
+  // The collection ID (or collection group) of the query target.
+  private readonly collectionId: string;
+  // The single inequality filter of the target (if it exists).
+  private readonly inequalityFilter?: FieldFilter;
+  // The list of equality filter of the target.
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/target_index_matcher.ts:60]: Done

File: packages/firestore/src/model/target_index_matcher.ts [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        !this.matchesFilter(this.inequalityFilter, segment) ||
+        !this.matchesOrderBy(this.orderBys[orderBysIndex++], segment)
+      ) {
+        return false;
+      }
+      ++segmentIndex;
+    }
+
+    // All remaining segment need to represent the prefix of the target's
+    // orderBy
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/target_index_matcher.ts:165]: Done

File: packages/firestore/test/unit/model/target_index_matcher.test.ts (hunk 2 above)
Comment (schmidt-sebastian) [reply packages/firestore/test/unit/model/target_index_matcher.test.ts:199]: Yes it applies to the assert.

**Review by schmidt-sebastian — COMMENTED** [review]
Wheee... so many typos. Thanks for catching.

**Comment by ehsannas** [comment]
forgot to push your commit?

**Comment by schmidt-sebastian** [comment]
@ehsannas Yup, had to pull and then push!

**Pull request merged**